File: web/src/domain/__tests__/HighlightClock.test.ts
import { describe, expect, it } from 'vitest'
import { HighlightClock } from '../HighlightClock'

describe('HighlightClock', () => {
  it('passes forward samples', () => {
    const clock = new HighlightClock()
    expect(clock.sample('a', 100)).toBe(100)
    expect(clock.sample('a', 133)).toBe(133)
    expect(clock.sample('a', 166)).toBe(166)
  })

  it('holds small backward corrections', () => {
    const clock = new HighlightClock()
    clock.sample('a', 1_000)
    expect(clock.sample('a', 960)).toBe(1_000)
    expect(clock.sample('a', 1_005)).toBe(1_005)
  })

  it('does not let repeated jitter creep backward', () => {
    const clock = new HighlightClock()
    clock.sample('a', 1_000)
    expect(clock.sample('a', 900)).toBe(1_000)
    expect(clock.sample('a', 850)).toBe(1_000)
    expect(clock.sample('a', 990)).toBe(1_000)
  })

  it('passes a genuine seek outside settle and resets for a new media item', () => {
    const clock = new HighlightClock(HighlightClock.SEEK_THRESHOLD_MS, 0)
    clock.sample('a', 5_000)
    expect(clock.sample('a', 1_000)).toBe(1_000)
    expect(clock.sample('b', 50)).toBe(50)
  })

  it('treats exactly the threshold as a seek outside settle', () => {
    const clock = new HighlightClock(250, 0)
    clock.sample('a', 1_000)
    expect(clock.sample('a', 750)).toBe(750)
  })

  it('acceptNextSample lets a short backward seek through', () => {
    const clock = new HighlightClock(250)
    clock.sample('a', 1_000)
    clock.acceptNextSample()
    expect(clock.sample('a', 900)).toBe(900)
    expect(clock.sample('a', 880)).toBe(900)
  })

  it('post-seek overshoot then snap-back does not bounce the clock', () => {
    const clock = new HighlightClock()
    clock.acceptNextSample()
    expect(clock.sample('ayah 4', 0)).toBe(0)
    expect(clock.sample('ayah 4', 800)).toBe(0)
    expect(clock.sample('ayah 4', 100)).toBe(100)
    expect(clock.sample('ayah 4', 40)).toBe(100)
    expect(clock.sample('ayah 4', 140)).toBe(140)
  })

  it('settle holds large regressions that would otherwise count as seeks', () => {
    const clock = new HighlightClock()
    clock.sample('a', 0)
    expect(clock.sample('a', 100)).toBe(100)
    expect(clock.sample('a', 200)).toBe(200)
    expect(clock.sample('a', 300)).toBe(300)
    expect(clock.sample('a', 400)).toBe(400)
    expect(clock.sample('a', 500)).toBe(500)
    expect(clock.sample('a', 100)).toBe(500)
  })
})

File: app/src/test/java/com/beautifulquran/domain/HighlightClockTest.kt
package com.beautifulquran.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class HighlightClockTest {

    @Test
    fun `forward samples pass through`() {
        val clock = HighlightClock()
        assertEquals(100L, clock.sample("a", 100))
        assertEquals(133L, clock.sample("a", 133))
        assertEquals(166L, clock.sample("a", 166))
    }

    @Test
    fun `small backward step is held at the previous position`() {
        val clock = HighlightClock()
        clock.sample("a", 1000)
        // MediaController extrapolation corrected backward by 40 ms — the
        // jitter that bounced the active word and flickered the ink.
        assertEquals(1000L, clock.sample("a", 960))
    }

    @Test
    fun `clock resumes from the raw position once it moves forward again`() {
        val clock = HighlightClock()
        clock.sample("a", 1000)
        clock.sample("a", 960)
        assertEquals(1005L, clock.sample("a", 1005))
    }

    @Test
    fun `repeated jitter cannot creep the clock backward`() {
        val clock = HighlightClock()
        clock.sample("a", 1000)
        assertEquals(1000L, clock.sample("a", 900))
        assertEquals(1000L, clock.sample("a", 850))
        assertEquals(1000L, clock.sample("a", 990))
    }

    @Test
    fun `a genuine backward seek passes through outside settle`() {
        val clock = HighlightClock(settleSamples = 0)
        clock.sample("a", 5000)
        // Word tap / loop restart: jump well past the jitter threshold.
        assertEquals(1000L, clock.sample("a", 1000))
        assertEquals(1033L, clock.sample("a", 1033))
    }

    @Test
    fun `a key change resets the clock even to an earlier position`() {
        val clock = HighlightClock()
        clock.sample("ayah 1", 45_000)
        // Ayah handoff: the next clip starts near zero.
        assertEquals(50L, clock.sample("ayah 2", 50))
    }

    @Test
    fun `regression exactly at the threshold is a seek outside settle`() {
        val clock = HighlightClock(seekThresholdMs = 250, settleSamples = 0)
        clock.sample("a", 1000)
        assertEquals(750L, clock.sample("a", 750))
    }

    @Test
    fun `regression just under the threshold is jitter`() {
        val clock = HighlightClock(seekThresholdMs = 250, settleSamples = 0)
        clock.sample("a", 1000)
        assertEquals(1000L, clock.sample("a", 751))
    }

    @Test
    fun `acceptNextSample lets a short backward seek through`() {
        val clock = HighlightClock(seekThresholdMs = 250)
        clock.sample("a", 1000)
        // Word tap 100 ms earlier would normally be held as jitter.
        clock.acceptNextSample()
        assertEquals(900L, clock.sample("a", 900))
        // Subsequent small regressions are held again.
        assertEquals(900L, clock.sample("a", 880))
    }

    @Test
    fun `post-seek overshoot then snap-back does not bounce the clock`() {
        // Back-button / previous-ayah: first sample is 0, then the controller
        // briefly reports ~800 ms (word 2–3), then a real ~100 ms, then a
        // correction back. Without settle that lights word 2, starts its wash,
        // then re-enters it.
        val clock = HighlightClock()
        clock.acceptNextSample()
        assertEquals(0L, clock.sample("ayah 4", 0))
        assertEquals(0L, clock.sample("ayah 4", 800)) // overshoot ignored
        assertEquals(100L, clock.sample("ayah 4", 100)) // believable advance
        assertEquals(100L, clock.sample("ayah 4", 40)) // snap-back held
        assertEquals(140L, clock.sample("ayah 4", 140))
    }

    @Test
    fun `settle holds large regressions that would otherwise count as seeks`() {
        val clock = HighlightClock(seekThresholdMs = 250)
        clock.sample("a", 0)
        // Ramp within the settle step cap so we are mid-settle at ~500 ms.
        assertEquals(100L, clock.sample("a", 100))
        assertEquals(200L, clock.sample("a", 200))
        assertEquals(300L, clock.sample("a", 300))
        assertEquals(400L, clock.sample("a", 400))
        assertEquals(500L, clock.sample("a", 500))
        // 400 ms regression is above SEEK_THRESHOLD but still in settle → hold.
        assertEquals(500L, clock.sample("a", 100))
    }

    @Test
    fun `after settle expires a large regression is accepted as a seek`() {
        val clock = HighlightClock(settleSamples = 2, maxSettleStepMs = 10_000)
        clock.sample("a", 5000)
        clock.sample("a", 5033) // settleLeft 1 → 0
        clock.sample("a", 5066) // settle exhausted
        assertEquals(1000L, clock.sample("a", 1000))
    }
}
